examples/d3d12/src/main.rs

```diff
@@ -0,0 +1,844 @@
+            command_list.IASetPrimitiveTopology(D3D_PRIMITIVE_TOPOLOGY_TRIANGLELIST);
+            command_list.IASetVertexBuffers(0, 1, &resources.vbv);
+            command_list.DrawInstanced(3, 1, 0, 0);
+
+            // Indicate that the back buffer will now be used to present.
+            command_list.ResourceBarrier(
+                1,
+                &transition_barrier(
+                    &resources.render_targets[resources.frame_index as usize],
+                    D3D12_RESOURCE_STATE_RENDER_TARGET,
+                    D3D12_RESOURCE_STATE_PRESENT,
+                ),
+            );
+        }
+
+        unsafe { command_list.Close() }.ok()
+    }
+
+    fn transition_barrier(
+        resource: &ID3D12Resource,
+        state_before: D3D12_RESOURCE_STATES,
+        state_after: D3D12_RESOURCE_STATES,
+    ) -> D3D12_RESOURCE_BARRIER {
+        D3D12_RESOURCE_BARRIER {
+            Type: D3D12_RESOURCE_BARRIER_TYPE_TRANSITION,
+            Flags: D3D12_RESOURCE_BARRIER_FLAG_NONE,
+            Anonymous: D3D12_RESOURCE_BARRIER_0 {
+                Transition: D3D12_RESOURCE_TRANSITION_BARRIER {
+                    pResource: Some(resource.clone()),
+                    StateBefore: state_before,
+                    StateAfter: state_after,
+                    Subresource: D3D12_RESOURCE_BARRIER_ALL_SUBRESOURCES,
+                }
+                .abi(),
+            },
+        }
+    }
+
+    fn create_device(command_line: &SampleCommandLine) -> Result<(IDXGIFactory4, ID3D12Device)> {
+        if cfg!(debug_assertions) {
+            unsafe {
+                if let Ok(debug) = D3D12GetDebugInterface::<ID3D12Debug>() {
+                    debug.EnableDebugLayer();
+                }
+            }
+        }
+
+        let dxgi_factory_flags = if cfg!(debug_assertions) {
+            DXGI_CREATE_FACTORY_DEBUG
+        } else {
+            0
+        };
+
+        let dxgi_factory: IDXGIFactory4 = unsafe { CreateDXGIFactory2(dxgi_factory_flags) }?;
+
+        let adapter = if command_line.use_warp_device {
+            unsafe { dxgi_factory.EnumWarpAdapter() }
+        } else {
+            get_hardware_adapter(&dxgi_factory)
+        }?;
+
+        let device = unsafe { D3D12CreateDevice(adapter, D3D_FEATURE_LEVEL_11_0) }?;
+        Ok((dxgi_factory, device))
+    }
+
+    fn create_root_signature(device: &ID3D12Device) -> Result<ID3D12RootSignature> {
+        let desc = D3D12_ROOT_SIGNATURE_DESC {
+            Flags: D3D12_ROOT_SIGNATURE_FLAG_ALLOW_INPUT_ASSEMBLER_INPUT_LAYOUT,
+            ..Default::default()
+        };
+
+        let mut error = None;
+        let mut signature = None;
+
+        let signature = unsafe {
+            D3D12SerializeRootSignature(
+                &desc,
+                D3D_ROOT_SIGNATURE_VERSION_1,
+                &mut signature,
+                &mut error,
+            )
+        }
+        .and_some(signature)?;
```

Comment:
That's a tricky one. I don't know if error-pointers for context as last argument are common (like they are in glib) but if they are it'd be nice if context could be added automagically in the bindings.

Also, would you need to check whether it's not `None` and panic instead? DirectXShaderCompiler's `Compile` method usually returns `S_OK` and then sets the `error` blob appropriately with a status code (not available on `ID3DBlob`) and extra error message in the data pointer+size.

Maybe this needs something of the form:
```suggestion
                &mut error,
            )
        };

        if let Some(error) = error {
            panic!(
                "Root signature serialization failed: {}",
                std::str::from_utf8(unsafe {
                    std::slice::from_raw_parts(
                        error.GetBufferPointer() as *const _,
                        error.GetBufferSize(),
                    )
                })
                .unwrap()
            );
        }
        let signature = hr.and_some(signature)?;
```

Reply:
Hey, thanks for reviewing this so carefully :)  It's not _hugely_ common for these APIs to include error output like this.  It it is optional, so I'll just make it null for the purposes of this example.

Reply:
Yes that's true, I was writing this as if it was for the `compile` function where the log is interesting to debug typos etc. Maybe worth adding it over there for that reason though?